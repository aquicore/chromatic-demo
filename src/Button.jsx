import PropTypes from "prop-types";
import React, { useCallback, useState } from "react";
import "./button.css";

// Utility Function
export const randomUtilityFunction = (a, b) => {
  return a + b;
}

// Custom Hook
const useOnClick = (handler) => {
  const [clicked, setClicked] = useState(false);

  const onClick = useCallback(() => {
    setClicked(true);
    handler();
  }, [handler]);

  return { clicked, onClick };
};

/**
 * Primary UI component for user interaction
 */
export const Button = ({ primary, backgroundColor, size, label, clickedText, ...props }) => {
  const mode = primary ? "storybook-button--primary" : "storybook-button--secondary";

  const { clicked, onClick } = useOnClick(props.onClick);

  const style = {
    border: "thin solid orange",
    backgroundColor,
  };

  return (
    <>
      <button
        type="button"
        className={["storybook-button", `storybook-button--${size}`, mode].join(" ")}
        style={style}
        {...props}
        onClick={onClick}
      >
        {label}
      </button>
      {clicked && <p>{clickedText}</p>}
    </>
  );
};

Button.propTypes = {
  /**
   * Is this the principal call to action on the page?
   */
  primary: PropTypes.bool,
  /**
   * What background color to use
   */
  backgroundColor: PropTypes.string,
  /**
   * How large should the button be?
   */
  size: PropTypes.oneOf(["small", "medium", "large"]),
  /**
   * Button contents
   */
  label: PropTypes.string.isRequired,
  /**
   * Optional click handler
   */
  onClick: PropTypes.func,
};

Button.defaultProps = {
  backgroundColor: null,
  primary: false,
  size: "medium",
  onClick: undefined,
};
