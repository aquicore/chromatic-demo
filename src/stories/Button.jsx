import PropTypes from "prop-types";
import React, { useCallback, useState } from "react";
import "./button.css";

/**
 * Primary UI component for user interaction
 */
export const Button = ({ primary, backgroundColor, size, label, onClick, clickedText, ...props }) => {
  const mode = primary ? "storybook-button--primary" : "storybook-button--secondary";

  const [clicked, setClicked] = useState(false);

  const internalOnClick = useCallback(() => {
    setClicked(true);
    onClick();
  }, [onClick]);

  const style = {
    border: "thin solid red",
    backgroundColor,
  };

  return (
    <>
      <button
        type="button"
        className={["storybook-button", `storybook-button--${size}`, mode].join(" ")}
        style={style}
        {...props}
        onClick={internalOnClick}
      >
        {label}
      </button>
      {clicked && <p>{clickedText}</p>}
    </>
  );
};

Button.propTypes = {
  /**
   * Is this the principal call to action on the page?
   */
  primary: PropTypes.bool,
  /**
   * What background color to use
   */
  backgroundColor: PropTypes.string,
  /**
   * How large should the button be?
   */
  size: PropTypes.oneOf(["small", "medium", "large"]),
  /**
   * Button contents
   */
  label: PropTypes.string.isRequired,
  /**
   * Optional click handler
   */
  onClick: PropTypes.func,
};

Button.defaultProps = {
  backgroundColor: null,
  primary: false,
  size: "medium",
  onClick: undefined,
};
