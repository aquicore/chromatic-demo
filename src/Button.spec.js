import { render } from "@testing-library/react";
import { Button, randomUtilityFunction } from "./Button";

describe("Button", () => {
    // This test is superfluous as it's already testing in Chromatic
    it("renders correctly", () => {
        const { asFragment } = render(<Button primary={true} backgroundColor="steelblue" label="Test Button" />);

        expect(asFragment()).toMatchSnapshot();
    })

    it("randomUtilityFunction sums correctly", () => {
        expect(randomUtilityFunction(3, 4)).toBe(7);
    })

})
