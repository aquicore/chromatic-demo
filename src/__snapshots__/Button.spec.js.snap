// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<DocumentFragment>
  <button
    class="storybook-button storybook-button--medium storybook-button--primary"
    style="border: thin solid orange; background-color: steelblue;"
    type="button"
  >
    Test Button
  </button>
</DocumentFragment>
`;
